src/plugins/gui/providers/moonpayProvider.ts

```diff
@@ -356,27 +384,194 @@ export const moonpayProvider: FiatProviderFactory = {
+              const queryObj: MoonpayBuyWidgetQueryParams = {
+                apiKey,
+                walletAddress: receiveAddress.publicAddress,
+                currencyCode: cryptoCurrencyObj.code,
+                paymentMethod,
+                baseCurrencyCode: fiatCurrencyObj.code,
+                lockAmount: true,
+                showAllCurrencies: false,
+                enableRecurringBuys: false
+              }
+              if (params.amountType === 'crypto') {
+                queryObj.quoteCurrencyAmount = moonpayQuote.quoteCurrencyAmount
+              } else {
+                queryObj.baseCurrencyAmount = 'totalAmount' in moonpayQuote ? moonpayQuote.totalAmount : undefined
+              }
+              urlObj.set('query', queryObj)
+              console.log('Approving moonpay buy quote url=' + urlObj.href)
+              await showUi.openExternalWebView({ url: urlObj.href })
             } else {
-              queryObj.baseCurrencyAmount = 'totalAmount' in moonpayQuote ? moonpayQuote.totalAmount : undefined
+              const urlObj = new URL('https://sell.moonpay.com?', true)
+              const queryObj: MoonpaySellWidgetQueryParams = {
+                apiKey,
+                refundWalletAddress: receiveAddress.publicAddress,
+                quoteCurrencyCode: fiatCurrencyObj.code,
+                paymentMethod,
+                baseCurrencyCode: cryptoCurrencyObj.code,
+                lockAmount: true,
+                showAllCurrencies: false,
+                redirectURL: RETURN_URL_PAYMENT
+              }
+              if (params.amountType === 'crypto') {
+                queryObj.baseCurrencyAmount = moonpayQuote.baseCurrencyAmount
+              } else {
+                queryObj.quoteCurrencyAmount = 'totalAmount' in moonpayQuote ? moonpayQuote.totalAmount : undefined
+              }
+              urlObj.set('query', queryObj)
+              console.log('Approving moonpay sell quote url=' + urlObj.href)
+
+              let inPayment = false
+
+              const openWebView = async () => {
+                await showUi.openWebView({
+                  url: urlObj.href,
+                  onUrlChange: async (uri: string) => {
+                    console.log('Moonpay WebView url change: ' + uri)
+
+                    if (uri.startsWith(RETURN_URL_PAYMENT)) {
+                      console.log('Moonpay WebView launch payment: ' + uri)
+                      const urlObj = new URL(uri, true)
+                      const { query } = urlObj
+                      const { baseCurrencyAmount, baseCurrencyCode, depositWalletAddress, depositWalletAddressTag, transactionId } = query
+                      if (inPayment) return
+                      inPayment = true
+                      try {
+                        if (baseCurrencyAmount == null || baseCurrencyCode == null || depositWalletAddress == null || transactionId == null) {
```

Comment:
Are we in control over these query parameters? Or is this something we expect moonpay to provide during the URL change event?

Reply:
moonpay provides these values. we have no control.